Lib/test/test_remote_pdb.py

```diff
@@ -78,6 +79,697 @@ def get_output(self) -> List[dict]:
         return results
 
 
+class MockDebuggerSocket:
+    """Mock file-like simulating a connection to a _RemotePdb instance"""
+
+    def __init__(self, incoming):
+        self.incoming = iter(incoming)
+        self.outgoing = []
+        self.buffered = bytearray()
+
+    def write(self, data: bytes) -> None:
+        """Simulate write to socket."""
+        self.buffered += data
+
+    def flush(self) -> None:
+        """Ensure each line is valid JSON."""
+        lines = self.buffered.splitlines(keepends=True)
+        self.buffered.clear()
+        for line in lines:
+            assert line.endswith(b"\n")
+            self.outgoing.append(json.loads(line))
+
+    def readline(self) -> bytes:
+        """Read a line from the prepared input queue."""
+        # Anything written must be flushed before trying to read,
+        # since the read will be dependent upon the last write.
+        assert not self.buffered
+        try:
+            item = next(self.incoming)
+            if not isinstance(item, bytes):
+                item = json.dumps(item).encode()
+            return item + b"\n"
+        except StopIteration:
+            return b""
+
+    def close(self) -> None:
+        """No-op close implementation."""
+        pass
+
+
+class PdbClientTestCase(unittest.TestCase):
+    """Tests for the _PdbClient class."""
+
+    def do_test(
+        self,
+        *,
+        incoming,
+        simulate_write_failure=False,
+        simulate_flush_failure=False,
+        expected_outgoing=[],
+        expected_completions=[],
+        expected_exception=None,
+        expected_stdout="",
+        expected_stdout_substring="",
```

Comment:
Do you think having both exact equal and partial equal important for our test cases? Can we only use `assertIn` or that does not give us enough coverage?

Reply:
I'd rather have both. In most cases we want to check an exact match, and there's only a handful of places where we want a substring - mostly to not make assumptions about the exact text of an exception. If you'd rather only have one, then I'd rather get rid of the substring search and only do the exact match, but that may mean higher maintenance costs to keep up with changes to other parts of CPython.

Reply:
> Can we only use `assertIn` or that does not give us enough coverage?

And specifically, that would miss cases where our output has extra newlines, or something like that. An easy mistake to make, considering one might come from the server and one might be added by the client.

Reply:
We can have both for now. I think it's going to be more flexible if it's a return value, but we do have multiple outputs (completion, outgoing messages, stdout) so maybe just keep it as it is and see if we need to change it in the future.